state/cloudcredentials.go

```diff
@@ -477,6 +477,39 @@ func (st *State) CredentialModels(tag names.CloudCredentialTag) (map[string]stri
 	return results, nil
 }
 
+// RemoveModelsCredential clears out given credential reference from all models that have it.
+func (st *State) RemoveModelsCredential(tag names.CloudCredentialTag) error {
+	coll, cleanup := st.db().GetCollection(modelsC)
+	defer cleanup()
+
+	sel := bson.D{
+		{"cloud-credential", tag.Id()},
+		{"life", bson.D{{"$ne", Dead}}},
+	}
+	iter := coll.Find(sel).Iter()
+	defer iter.Close()
+
+	var ops []txn.Op
+	var doc bson.M
+	for iter.Next(&doc) {
+		id, ok := doc["_id"]
+		if !ok {
+			return errors.New("no id found in model doc")
+		}
+
+		ops = append(ops, txn.Op{
+			C:      modelsC,
+			Id:     id,
+			Assert: txn.DocExists,
```

Comment:
Need to assert $ne Dead here too based on the initial query. Also, if the model is deleted before the txn runs, it will error out, so need to include in a buildTxn loop. Either that or remove the DocExists assert since if the model is gone, who cares that the credential cannot be cleared.

Based on this issue, there should be a test for such a race condition using BeforeHook